lib/std/Target.zig

```diff
@@ -1147,6 +1153,8 @@ pub const Cpu = struct {
         powerpcle,
         powerpc64,
         powerpc64le,
+        propeller1,
+        propeller2,
```

Comment:
> I've thought about this a bit more and I think I'd prefer representing P2 as a CPU feature. My reasoning is that we actually have precedent for this already; for example, MIPS r6 made many backwards-incompatible changes (e.g. reallocated encodings, removed instructions), yet we don't represent that as a distinct Arch tag. I believe PowerPC has also historically removed features, but I can't remember the exact details there.

This would mean we have no `generic` cpu then, but only `propeller1` and `propeller2`? They are neither binary compatible nor instruction compatible, it's more like arm vs. aarch64 than arm vs. thumb

This would also mean you can't check on arch but always have to check on the actual CPU in your code as well, so if you ever need to do something for p1 and p2, it would look like this:
```zig
switch(builtin.cpu.arch) {
  .propeller => switch(builtin.cpu) {
    &std.Target.propeller.cpus.propeller1 => { ... },
    &std.Target.propeller.cpus.propeller1 => { ... },
  },
  ...
}
```

I wonder why they made them a CPU/Feature in the gcc port. Probably due to convenience as otherwise you have to use `prop1-none-eabi-gcc` and `prop2-none-eabi-gcc` instead of `prop-none-eabi-gcc`

Reply:
> This would mean we have no `generic` cpu then, but only `propeller1` and `propeller2`? They are neither binary compatible nor instruction compatible, it's more like arm vs. aarch64 than arm vs. thumb

Yes; this is no different from the MIPS situation. The "generic" CPU would be arbitrarily defined to be `propeller1`, just as `mips32`/`mips64` (r1) are arbitrarily defined to be the generic CPUs for MIPS.

> This would also mean you can't check on arch but always have to check on the actual CPU

You would check the feature flag that the CPU would imply; specifically, `propeller2` would imply `p2`. That currently looks like `std.Target.propeller.featureSetHas(target.cpu.features, .p2)` which is indeed too verbose, but I have some ideas for making this API more ergonomic that I plan to get to soon. I think something like `target.cpu.features.has(.propeller, .p2)` should be possible with some light comptime magic.

> I wonder why they made them a CPU/Feature in the gcc port.

From what I saw in GCC and binutils, there's still a decent amount of overlap in instruction encodings, and they don't use wholly separate en/decoding tables or anything. In this way, I don't think the arm vs aarch64 comparison quite works because A64 really is an entirely new instruction set from A32.

Reply:
Let's wait until we have all toolchains tested if they agree on the C type sizes

Reply:
I think one heuristic to determine whether it should be a CPU feature vs a different arch is whether you would want two different self-hosted backend implementations, or whether you would rather have one that checks the CPU feature. If they would share a lot of implementation code, that is a hint that the same CPU arch would make sense.

Reply:
> I think one heuristic to determine whether it should be a CPU feature vs a different arch is whether you would want two different self-hosted backend implementations, or whether you would rather have one that checks the CPU feature. If they would share a lot of implementation code, that is a hint that the same CPU arch would make sense.

That's a pretty good heuristic and i think i have to take a deep-dive into codegen for that :grin: 

They both have basically the same *idea* behind the instruction encoding and style, but the semantics of basically all except the most basic instructions are different.

It starts with `RDLONG` which can only read Hub memory on P1 and reads continuous memory on P2.

`@sizeOf(usize)` would be 32 bit with 23 bit padding for 2 of the 3 address spaces so it's kinda hard to say.

But now i'm super tempted to create a self-hosted backend for the two processors... :thinking: 

Next steps for me is still getting all the compilers up and running, which sadly takes a lot of time :(